File: ExtenderSample_vb/ControlExtender.vb
Imports System.ComponentModel
Imports System.Windows.Forms

<ProvideProperty("SomeProperty", GetType(Control))>
Public Class ControlExtender
    Inherits Component
    Implements IExtenderProvider
    Private controls As New Hashtable
    Public Function CanExtend(extendee As Object) As Boolean Implements IExtenderProvider.CanExtend
        Return TypeOf extendee Is Control
    End Function

    Public Function GetSomeProperty(control As Control) As String
        If controls.ContainsKey(control) Then
            Return DirectCast(controls(control), String)
        End If

        Return Nothing
    End Function
    Public Sub SetSomeProperty(control As Control, value As String)
        If (String.IsNullOrEmpty(value)) Then
            controls.Remove(control)
        Else
            controls(control) = value
        End If
    End Sub
End Class

File: ExtenderSample/ControlExtender.cs
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

namespace ExtenderSample
{
    [ProvideProperty("Custom1", typeof(Control))]
    [ProvideProperty("Custom2", typeof(Control))]
    public class ControlExtender : Component, IExtenderProvider
    {
        private Hashtable _controls = new Hashtable();
        public bool CanExtend(object extendee)
        {
            return extendee is Control;
        }

        public string GetCustom1(Control control)
        {
            if (_controls.ContainsKey(control))
            {
                return (string)_controls[control];
            }

            return null;
        }
        public string GetCustom2(Control control)
        {
            if (_controls.ContainsKey(control))
            {
                return (string)_controls[control];
            }

            return null;
        }
        public void SetCustom1(Control control, string value)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                _controls.Remove(control);
            }
            else
            {
                _controls[control] = value;
            }
        }
        public void SetCustom2(Control control, int? value)
        {
            if (!value.HasValue)
            {
                _controls.Remove(control);
            }
            else
            {
                _controls[control] = value;
            }
        }
    }

}
